core/src/main/java/org/apache/brooklyn/core/objs/BasicSpecParameter.java

```diff
@@ -404,7 +431,126 @@ public static void addParameters(AbstractBrooklynObjectSpec<?, ?> spec, List<? e
+                existingToKeep.put(p.getConfigKey().getName(), p);
+            }
+        }
+
+        List<SpecParameter<?>> result = MutableList.<SpecParameter<?>>of();
+        for (SpecParameter<?> p: newParams) {
+            if (p instanceof SpecParameterForInheritance) {
+                SpecParameter<?> existingP = existingToKeep.remove(p.getConfigKey().getName());
+                if (existingP!=null) {
+                    p = ((SpecParameterForInheritance<?>)p).resolveWithAncestor(existingP);
+                } else {
+                    // TODO find config keys on the type (not available as parameters)
+                    /* we don't currently do this due to low priority; all it means if there is a config key in java,
+                     * and a user wishes to expose it as a parameter, they have to redeclare everything;
+                     * nothing from the config key in java will be inherited */
+                    p = ((SpecParameterForInheritance<?>)p).resolveWithAncestor((ConfigKey<?>)null);
+                }
+                result.add(p);
+            } else {
+                existingToKeep.remove(p.getConfigKey().getName());
+                result.add(p);
+            }
+        }
+        result.addAll(existingToKeep.values());
+        return result;
+    }
+
+    /** instance of {@link SpecParameter} which includes information about what fields are explicitly set,
+     * for use with a subsequent merge */
+    @SuppressWarnings("serial")
+    @Beta
+    static class SpecParameterForInheritance<T> extends BasicSpecParameter<T> {
+
+        private final boolean hasType, hasLabelSet, hasPinnedSet, hasDefaultValue, hasConstraints, hasRuntimeInheritance, hasTypeInheritance;
+
+        private <SensorType> SpecParameterForInheritance(String label, Boolean pinned, ConfigKey<T> config, AttributeSensor<SensorType> sensor,
+                boolean hasType, boolean hasDefaultValue, boolean hasConstraints, boolean hasRuntimeInheritance, boolean hasTypeInheritance) {
+            super(Preconditions.checkNotNull(label!=null ? label : config.getName(), "label or config name must be set"), 
+                    pinned==null ? true : pinned, config, sensor);
+            this.hasType = hasType;
+            this.hasLabelSet = label!=null;
+            this.hasPinnedSet = pinned!=null;
+            this.hasDefaultValue = hasDefaultValue;
+            this.hasConstraints = hasConstraints;
+            this.hasRuntimeInheritance = hasRuntimeInheritance;
+            this.hasTypeInheritance = hasTypeInheritance;
+        }
+
+        /** used if yaml specifies a parameter, but possibly incomplete, and a spec supertype has a parameter */
+        SpecParameter<?> resolveWithAncestor(SpecParameter<?> ancestor) {
+            if (ancestor==null) return new BasicSpecParameter<>(getLabel(), isPinned(), getConfigKey(), getSensor());
+
+            return new BasicSpecParameter<>(
```

Comment:
yes, should always be top-down, and the `SpecParameterForInheritance` items replaced immediately.  i'll have another look.